File: internal/restrictionprediction/parse.go
package restrictionprediction

import (
	"bytes"
	"encoding/json"
	"errors"

	"github.com/blisspixel/fartapp/internal/restrictionflow"
	"github.com/blisspixel/fartapp/internal/strictjson"
)

type requestDocument struct {
	Schema               string            `json:"schema"`
	Model                requestModel      `json:"model"`
	QuantitySystem       string            `json:"quantity_system"`
	Stagnation           requestStagnation `json:"stagnation"`
	BackPressurePascals  *float64          `json:"back_pressure_pa"`
	DischargeCoefficient *float64          `json:"discharge_coefficient"`
	Area                 requestArea       `json:"area"`
}

type requestModel struct {
	ID      string `json:"id"`
	Version string `json:"version"`
}

type requestStagnation struct {
	PressurePascals     *float64 `json:"pressure_pa"`
	TemperatureKelvin   *float64 `json:"temperature_k"`
	SpecificGasConstant *float64 `json:"specific_gas_constant_j_per_kg_k"`
	HeatCapacityRatio   *float64 `json:"heat_capacity_ratio"`
}

type requestArea struct {
	Law                    string   `json:"law"`
	PrescribedSquareMetres *float64 `json:"prescribed_m2"`
	ComplianceSquareMetres *float64 `json:"compliance_m2_per_pa"`
	MaximumSquareMetres    *float64 `json:"maximum_m2"`
}

type parsedRequest struct {
	request restrictionflow.Request
}

func parseRequest(data []byte) (parsedRequest, *Diagnostic) {
	if diagnostic := preflightJSON(data); diagnostic != nil {
		return parsedRequest{}, diagnostic
	}
	if issue := strictjson.InspectShape[requestDocument](data); issue != nil {
		return parsedRequest{}, schemaDiagnostic(issue.Path, "document_shape_invalid")
	}
	decoder := json.NewDecoder(bytes.NewReader(data))
	decoder.DisallowUnknownFields()
	var document requestDocument
	if err := decoder.Decode(&document); err != nil {
		return parsedRequest{}, schemaDiagnostic("/", "document_shape_invalid")
	}
	if document.Schema != RequestSchema {
		return parsedRequest{}, schemaDiagnostic("/schema", "unsupported_schema")
	}
	if document.Model.ID != ModelID || document.Model.Version != ModelVersion {
		return parsedRequest{}, schemaDiagnostic("/model", "unsupported_model_revision")
	}
	if document.QuantitySystem != QuantitySystem {
		return parsedRequest{}, schemaDiagnostic("/quantity_system", "unsupported_quantity_system")
	}
	if document.BackPressurePascals == nil {
		return parsedRequest{}, schemaDiagnostic("/back_pressure_pa", "missing_member")
	}
	if document.DischargeCoefficient == nil {
		return parsedRequest{}, schemaDiagnostic("/discharge_coefficient", "missing_member")
	}
	stagnation, diagnostic := parseStagnation(document.Stagnation)
	if diagnostic != nil {
		return parsedRequest{}, diagnostic
	}
	area, diagnostic := parseArea(document.Area)
	if diagnostic != nil {
		return parsedRequest{}, diagnostic
	}
	back, err := restrictionflow.NewPressure(*document.BackPressurePascals)
	if err != nil {
		return parsedRequest{}, modelDiagnostic("/back_pressure_pa", classifyModelError(err))
	}
	cd, err := restrictionflow.NewDischargeCoefficient(*document.DischargeCoefficient)
	if err != nil {
		return parsedRequest{}, modelDiagnostic("/discharge_coefficient", classifyModelError(err))
	}
	request, err := restrictionflow.NewRequest(stagnation, back, area, cd)
	if err != nil {
		return parsedRequest{}, modelDiagnostic("/", classifyModelError(err))
	}
	return parsedRequest{request: request}, nil
}

func parseStagnation(document requestStagnation) (restrictionflow.Stagnation, *Diagnostic) {
	if document.PressurePascals == nil {
		return restrictionflow.Stagnation{}, schemaDiagnostic("/stagnation/pressure_pa", "missing_member")
	}
	if document.TemperatureKelvin == nil {
		return restrictionflow.Stagnation{}, schemaDiagnostic("/stagnation/temperature_k", "missing_member")
	}
	if document.SpecificGasConstant == nil {
		return restrictionflow.Stagnation{}, schemaDiagnostic("/stagnation/specific_gas_constant_j_per_kg_k", "missing_member")
	}
	if document.HeatCapacityRatio == nil {
		return restrictionflow.Stagnation{}, schemaDiagnostic("/stagnation/heat_capacity_ratio", "missing_member")
	}
	pressure, err := restrictionflow.NewPressure(*document.PressurePascals)
	if err != nil {
		return restrictionflow.Stagnation{}, modelDiagnostic("/stagnation/pressure_pa", classifyModelError(err))
	}
	temperature, err := restrictionflow.NewTemperature(*document.TemperatureKelvin)
	if err != nil {
		return restrictionflow.Stagnation{}, modelDiagnostic("/stagnation/temperature_k", classifyModelError(err))
	}
	gasConstant, err := restrictionflow.NewSpecificGasConstant(*document.SpecificGasConstant)
	if err != nil {
		return restrictionflow.Stagnation{}, modelDiagnostic(
			"/stagnation/specific_gas_constant_j_per_kg_k",
			classifyModelError(err),
		)
	}
	gamma, err := restrictionflow.NewHeatCapacityRatio(*document.HeatCapacityRatio)
	if err != nil {
		return restrictionflow.Stagnation{}, modelDiagnostic(
			"/stagnation/heat_capacity_ratio",
			classifyModelError(err),
		)
	}
	stagnation, err := restrictionflow.NewStagnation(pressure, temperature, gasConstant, gamma)
	if err != nil {
		return restrictionflow.Stagnation{}, modelDiagnostic("/stagnation", classifyModelError(err))
	}
	return stagnation, nil
}

func parseArea(document requestArea) (restrictionflow.AreaLaw, *Diagnostic) {
	if document.PrescribedSquareMetres == nil {
		return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/prescribed_m2", "missing_member")
	}
	prescribed, err := restrictionflow.NewArea(*document.PrescribedSquareMetres)
	if err != nil {
		return restrictionflow.AreaLaw{}, modelDiagnostic("/area/prescribed_m2", classifyModelError(err))
	}
	switch document.Law {
	case "prescribed":
		if document.ComplianceSquareMetres != nil {
			return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/compliance_m2_per_pa", "unexpected_member")
		}
		if document.MaximumSquareMetres != nil {
			return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/maximum_m2", "unexpected_member")
		}
		law, err := restrictionflow.NewPrescribedArea(prescribed)
		if err != nil {
			return restrictionflow.AreaLaw{}, modelDiagnostic("/area", classifyModelError(err))
		}
		return law, nil
	case "linear-compliance":
		if document.ComplianceSquareMetres == nil {
			return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/compliance_m2_per_pa", "missing_member")
		}
		if document.MaximumSquareMetres == nil {
			return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/maximum_m2", "missing_member")
		}
		compliance, err := restrictionflow.NewAreaCompliance(*document.ComplianceSquareMetres)
		if err != nil {
			return restrictionflow.AreaLaw{}, modelDiagnostic("/area/compliance_m2_per_pa", classifyModelError(err))
		}
		maximum, err := restrictionflow.NewArea(*document.MaximumSquareMetres)
		if err != nil {
			return restrictionflow.AreaLaw{}, modelDiagnostic("/area/maximum_m2", classifyModelError(err))
		}
		law, err := restrictionflow.NewLinearComplianceArea(prescribed, compliance, maximum)
		if err != nil {
			return restrictionflow.AreaLaw{}, modelDiagnostic("/area", classifyModelError(err))
		}
		return law, nil
	default:
		return restrictionflow.AreaLaw{}, schemaDiagnostic("/area/law", "unsupported_area_law")
	}
}

func classifyModelError(err error) string {
	switch {
	case errors.Is(err, restrictionflow.ErrNonFiniteValue):
		return "nonfinite_quantity"
	case errors.Is(err, restrictionflow.ErrNonPositiveValue):
		return "nonpositive_quantity"
	case errors.Is(err, restrictionflow.ErrNegativeArea):
		return "negative_area"
	case errors.Is(err, restrictionflow.ErrNegativeCompliance):
		return "negative_compliance"
	case errors.Is(err, restrictionflow.ErrInvalidDischargeCoefficient):
		return "invalid_discharge_coefficient"
	case errors.Is(err, restrictionflow.ErrInvalidHeatCapacityRatio):
		return "invalid_heat_capacity_ratio"
	case errors.Is(err, restrictionflow.ErrInvalidAreaLaw):
		return "invalid_area_law"
	case errors.Is(err, restrictionflow.ErrInvalidStagnation):
		return "invalid_stagnation"
	case errors.Is(err, restrictionflow.ErrAdversePressure):
		return "adverse_pressure"
	case errors.Is(err, restrictionflow.ErrNoRepresentableFlow):
		return "no_representable_flow"
	default:
		return "numerical_domain_error"
	}
}

func schemaDiagnostic(path, reason string) *Diagnostic {
	return &Diagnostic{Code: "FART-E-SCHEMA-0003", Stage: "schema", Path: path, ReasonCode: reason}
}

func modelDiagnostic(path, reason string) *Diagnostic {
	return &Diagnostic{Code: "FART-E-MODEL-0003", Stage: "model", Path: path, ReasonCode: reason}
}

func preflightJSON(data []byte) *Diagnostic {
	issue := strictjson.Inspect(data, strictjson.Limits{
		MaximumDepth: maximumJSONDepth, MaximumMemberNameBytes: maximumMemberNameBytes,
	})
	if issue == nil {
		return nil
	}
	return &Diagnostic{
		Code: "FART-E-JSON-0003", Stage: "syntax", Path: issue.Path, ReasonCode: string(issue.Kind),
	}
}

File: crates/fart-services/src/restriction/parse.rs
use fart_domain::restriction::{
    Area, AreaCompliance, AreaLaw, DischargeCoefficient, HeatCapacityRatio, HistorySample,
    MAX_HISTORY_SAMPLES, Pressure, Request, Seconds, Stagnation,
};
use fart_domain::{SpecificGasConstant, Temperature};
use serde_json::Value;

use super::{HISTORY_REQUEST_SCHEMA, Kind, MODEL_ID, MODEL_VERSION, REQUEST_SCHEMA, wire};
use crate::Diagnostic;

pub(super) struct HistoryRequest {
    pub stagnation: Stagnation,
    pub back: Pressure,
    pub coefficient: DischargeCoefficient,
    pub samples: Vec<HistorySample>,
}

pub(super) fn prediction(data: &[u8]) -> Result<Request, Diagnostic> {
    let kind = Kind::Prediction;
    let document = wire::document(data, kind)?;
    header(&document, kind)?;
    let back = required(&document, "back_pressure_pa", "", kind)?;
    let coefficient = required(&document, "discharge_coefficient", "", kind)?;
    let stagnation = stagnation(&document["stagnation"], kind)?;
    let area = area(&document["area"])?;
    let back = Pressure::new(back)
        .map_err(|e| wire::issue(kind, "model", "/back_pressure_pa", e.reason()))?;
    let coefficient = DischargeCoefficient::new(coefficient)
        .map_err(|e| wire::issue(kind, "model", "/discharge_coefficient", e.reason()))?;
    Ok(Request::new(stagnation, back, area, coefficient))
}

pub(super) fn history(data: &[u8]) -> Result<HistoryRequest, Diagnostic> {
    let kind = Kind::History;
    let document = wire::document(data, kind)?;
    header(&document, kind)?;
    let back = required(&document, "back_pressure_pa", "", kind)?;
    let coefficient = required(&document, "discharge_coefficient", "", kind)?;
    let samples = document["samples"]
        .as_array()
        .filter(|values| !values.is_empty())
        .ok_or_else(|| wire::issue(kind, "schema", "/samples", "missing_member"))?;
    if samples.len() > MAX_HISTORY_SAMPLES {
        return Err(wire::issue(
            kind,
            "schema",
            "/samples",
            "invalid_sample_count",
        ));
    }
    let stagnation = stagnation(&document["stagnation"], kind)?;
    let back = Pressure::new(back)
        .map_err(|_| wire::issue(kind, "model", "/back_pressure_pa", "nonpositive_quantity"))?;
    let coefficient = DischargeCoefficient::new(coefficient).map_err(|_| {
        wire::issue(
            kind,
            "model",
            "/discharge_coefficient",
            "invalid_discharge_coefficient",
        )
    })?;
    let mut interpreted = Vec::with_capacity(samples.len());
    for (index, sample) in samples.iter().enumerate() {
        let path = format!("/samples/{index}");
        let (Some(time), Some(area)) =
            (sample["time_s"].as_f64(), sample["prescribed_m2"].as_f64())
        else {
            return Err(wire::issue(kind, "schema", &path, "missing_member"));
        };
        let time = Seconds::new(time)
            .map_err(|_| wire::issue(kind, "model", &format!("{path}/time_s"), "invalid_time"))?;
        let area = Area::new(area).map_err(|_| {
            wire::issue(
                kind,
                "model",
                &format!("{path}/prescribed_m2"),
                "negative_area",
            )
        })?;
        interpreted.push(HistorySample::new(time, area));
    }
    Ok(HistoryRequest {
        stagnation,
        back,
        coefficient,
        samples: interpreted,
    })
}

fn header(document: &Value, kind: Kind) -> Result<(), Diagnostic> {
    let (schema, path) = match kind {
        Kind::Prediction => (REQUEST_SCHEMA, "/schema"),
        Kind::History => (HISTORY_REQUEST_SCHEMA, "/"),
    };
    if document["schema"].as_str() != Some(schema) {
        return Err(wire::issue(kind, "schema", path, "unsupported_schema"));
    }
    if document["model"]["id"].as_str() != Some(MODEL_ID)
        || document["model"]["version"].as_str() != Some(MODEL_VERSION)
    {
        return Err(wire::issue(
            kind,
            "schema",
            "/model",
            "unsupported_model_revision",
        ));
    }
    if document["quantity_system"].as_str() != Some("si") {
        return Err(wire::issue(
            kind,
            "schema",
            "/quantity_system",
            "unsupported_quantity_system",
        ));
    }
    Ok(())
}

fn required(document: &Value, key: &str, base: &str, kind: Kind) -> Result<f64, Diagnostic> {
    document[key]
        .as_f64()
        .ok_or_else(|| wire::issue(kind, "schema", &format!("{base}/{key}"), "missing_member"))
}

fn stagnation(document: &Value, kind: Kind) -> Result<Stagnation, Diagnostic> {
    let fields = [
        "pressure_pa",
        "temperature_k",
        "specific_gas_constant_j_per_kg_k",
        "heat_capacity_ratio",
    ];
    let mut values = [0.0; 4];
    for (index, key) in fields.iter().enumerate() {
        values[index] = required(document, key, "/stagnation", kind).map_err(|error| {
            if kind == Kind::History {
                wire::issue(kind, "schema", "/stagnation", "missing_member")
            } else {
                error
            }
        })?;
    }
    let pressure = Pressure::new(values[0])
        .map_err(|e| wire::issue(kind, "model", "/stagnation/pressure_pa", e.reason()))?;
    let temperature = Temperature::new(values[1])
        .map_err(|e| wire::issue(kind, "model", "/stagnation/temperature_k", e.reason()))?;
    let gas_constant = SpecificGasConstant::new(values[2]).map_err(|e| {
        wire::issue(
            kind,
            "model",
            "/stagnation/specific_gas_constant_j_per_kg_k",
            e.reason(),
        )
    })?;
    let gamma = HeatCapacityRatio::new(values[3])
        .map_err(|e| wire::issue(kind, "model", "/stagnation/heat_capacity_ratio", e.reason()))?;
    Stagnation::new(pressure, temperature, gas_constant, gamma)
        .map_err(|e| wire::issue(kind, "model", "/stagnation", e.reason()))
}

fn area(document: &Value) -> Result<AreaLaw, Diagnostic> {
    let kind = Kind::Prediction;
    let prescribed = required(document, "prescribed_m2", "/area", kind)?;
    let prescribed = Area::new(prescribed)
        .map_err(|e| wire::issue(kind, "model", "/area/prescribed_m2", e.reason()))?;
    match document["law"].as_str().unwrap_or("") {
        "prescribed" => {
            for key in ["compliance_m2_per_pa", "maximum_m2"] {
                if document.get(key).is_some() {
                    return Err(wire::issue(
                        kind,
                        "schema",
                        &format!("/area/{key}"),
                        "unexpected_member",
                    ));
                }
            }
            Ok(AreaLaw::prescribed(prescribed))
        }
        "linear-compliance" => {
            let compliance = required(document, "compliance_m2_per_pa", "/area", kind)?;
            let maximum = required(document, "maximum_m2", "/area", kind)?;
            let compliance = AreaCompliance::new(compliance).map_err(|e| {
                wire::issue(kind, "model", "/area/compliance_m2_per_pa", e.reason())
            })?;
            let maximum = Area::new(maximum)
                .map_err(|e| wire::issue(kind, "model", "/area/maximum_m2", e.reason()))?;
            AreaLaw::linear_compliance(prescribed, compliance, maximum)
                .map_err(|e| wire::issue(kind, "model", "/area", e.reason()))
        }
        _ => Err(wire::issue(
            kind,
            "schema",
            "/area/law",
            "unsupported_area_law",
        )),
    }
}
